File: go/tests/security_watchdog_test.go
// security_watchdog_test.go
//
// Verifies DefaultSecurityWatchdog (ported from ISecurityWatchdog.cs):
//   - Graduated response policy across the confidence bands.
//   - High-severity vectors add StateRollback only when a VERIFYING checkpoint
//     is supplied.
//   - StreamSignals delivers every dispatched signal, buffers signals emitted
//     BEFORE a reader attaches (unbounded semantics), and completes on cancel.

package circleai_test

import (
	"context"
	"strings"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func sig(vec circleai.ThreatVector, confidence float32) *circleai.AnomalySignal {
	return circleai.NewAnomalySignal(vec, confidence, "CircleAI.Test", "unit signal", nil)
}

func TestWatchdog_LowConfidenceNoAction(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	resp, err := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.10), nil)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if resp.Kind != circleai.SecurityResponseKindNoAction {
		t.Errorf("kind: got %v, want NoAction", resp.Kind)
	}
	if !strings.Contains(resp.Description, "10%") {
		t.Errorf("description should show 10%%: %q", resp.Description)
	}
}

func TestWatchdog_MidConfidenceKeyRotation(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	resp, err := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.45), nil)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if resp.Kind != circleai.SecurityResponseKindKeyRotation {
		t.Errorf("kind: got %v, want KeyRotation", resp.Kind)
	}
	if !strings.Contains(resp.Description, "MemoryAnomaly") || !strings.Contains(resp.Description, "45%") {
		t.Errorf("description mismatch: %q", resp.Description)
	}
}

func TestWatchdog_BoundaryAtRotationThreshold(t *testing.T) {
	// Exactly 0.30 is NOT < 0.30, so it should rotate, not no-action.
	w := circleai.NewDefaultSecurityWatchdog()
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.30), nil)
	if resp.Kind != circleai.SecurityResponseKindKeyRotation {
		t.Errorf("0.30 should be KeyRotation (not < threshold), got %v", resp.Kind)
	}
}

func TestWatchdog_BoundaryBelowCompositeThreshold(t *testing.T) {
	// The composite gate is strict-greater (C#: confidence > 0.60). A value
	// clearly at or below the threshold stays KeyRotation. (Exactly 0.60 is a
	// float32 knife-edge: the Go port stores Confidence as float32 — matching the
	// existing schema — and float32(0.60) ≈ 0.60000002 which IS > 0.60, so the
	// exact-boundary result is representation-defined and not asserted here.)
	w := circleai.NewDefaultSecurityWatchdog()
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.59), nil)
	if resp.Kind != circleai.SecurityResponseKindKeyRotation {
		t.Errorf("0.59 should stay KeyRotation (≤ composite threshold), got %v", resp.Kind)
	}
}

func TestWatchdog_JustAboveCompositeThreshold(t *testing.T) {
	// Just above 0.60 escalates to Composite.
	w := circleai.NewDefaultSecurityWatchdog()
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.61), nil)
	if resp.Kind != circleai.SecurityResponseKindComposite {
		t.Errorf("0.61 should escalate to Composite (> threshold), got %v", resp.Kind)
	}
}

func TestWatchdog_HighConfidenceComposite(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.95), nil)
	if resp.Kind != circleai.SecurityResponseKindComposite {
		t.Fatalf("kind: got %v, want Composite", resp.Kind)
	}
	// Non-high-severity vector + no checkpoint → rotation + mesh only.
	if len(resp.AppliedActions) != 2 {
		t.Fatalf("expected 2 actions, got %v", resp.AppliedActions)
	}
	if resp.RestoredCheckpoint != nil {
		t.Error("no checkpoint supplied → RestoredCheckpoint must be nil")
	}
}

func TestWatchdog_HighSeverityWithVerifyingCheckpointAddsRollback(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	cp, _ := circleai.NewSecurityCheckpoint("u", "CircleAI.Memory", []byte("intact"))
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorStateCorruption, 0.99), cp)
	if resp.Kind != circleai.SecurityResponseKindComposite {
		t.Fatalf("kind: got %v", resp.Kind)
	}
	if len(resp.AppliedActions) != 3 {
		t.Fatalf("expected 3 actions (rotation+mesh+rollback), got %v", resp.AppliedActions)
	}
	if resp.AppliedActions[2] != circleai.SecurityResponseKindStateRollback {
		t.Errorf("third action should be StateRollback, got %v", resp.AppliedActions[2])
	}
	if resp.RestoredCheckpoint != cp {
		t.Error("RestoredCheckpoint should be the supplied checkpoint")
	}
}

func TestWatchdog_HighSeverityWithTamperedCheckpointSkipsRollback(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	cp, _ := circleai.NewSecurityCheckpoint("u", "CircleAI.Memory", []byte("intact"))
	cp.Payload[0] ^= 0xFF // fails Verify()
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorNetworkPivot, 0.99), cp)
	if len(resp.AppliedActions) != 2 {
		t.Fatalf("tampered checkpoint must NOT add rollback, got %v", resp.AppliedActions)
	}
	if resp.RestoredCheckpoint != nil {
		t.Error("tampered checkpoint → RestoredCheckpoint nil")
	}
}

func TestWatchdog_NonHighSeverityWithCheckpointSkipsRollback(t *testing.T) {
	// MemoryAnomaly is NOT in the high-severity set, so even a valid checkpoint
	// is not restored.
	w := circleai.NewDefaultSecurityWatchdog()
	cp, _ := circleai.NewSecurityCheckpoint("u", "CircleAI.Memory", []byte("intact"))
	resp, _ := w.OnAnomalyDetected(context.Background(), sig(circleai.ThreatVectorMemoryAnomaly, 0.99), cp)
	if len(resp.AppliedActions) != 2 {
		t.Fatalf("non-high-severity must NOT add rollback, got %v", resp.AppliedActions)
	}
}

func TestWatchdog_NilSignalErrors(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	if _, err := w.OnAnomalyDetected(context.Background(), nil, nil); err == nil {
		t.Error("nil signal should error")
	}
}

func TestWatchdog_StreamDeliversSignals(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Subscribe SYNCHRONOUSLY before emitting, then read on a goroutine.
	stream := w.StreamSignals(ctx)

	const n = 5
	go func() {
		for i := 0; i < n; i++ {
			_, _ = w.OnAnomalyDetected(ctx, sig(circleai.ThreatVectorMemoryAnomaly, 0.5), nil)
		}
	}()

	got := 0
	deadline := time.After(2 * time.Second)
	for got < n {
		select {
		case _, ok := <-stream:
			if !ok {
				t.Fatalf("stream closed early after %d signals", got)
			}
			got++
		case <-deadline:
			t.Fatalf("timed out after %d/%d signals", got, n)
		}
	}
}

func TestWatchdog_StreamBuffersPreSubscriptionSignals(t *testing.T) {
	// Unbounded semantics: signals emitted BEFORE ReadAll must still be seen.
	w := circleai.NewDefaultSecurityWatchdog()
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Emit two signals with NO active reader.
	_, _ = w.OnAnomalyDetected(ctx, sig(circleai.ThreatVectorMemoryAnomaly, 0.5), nil)
	_, _ = w.OnAnomalyDetected(ctx, sig(circleai.ThreatVectorControlFlowDrift, 0.5), nil)

	// Now attach a reader; both buffered signals must arrive.
	stream := w.StreamSignals(ctx)
	got := 0
	deadline := time.After(2 * time.Second)
	for got < 2 {
		select {
		case _, ok := <-stream:
			if !ok {
				t.Fatalf("stream closed after %d buffered signals", got)
			}
			got++
		case <-deadline:
			t.Fatalf("buffered signals lost: got %d/2", got)
		}
	}
}

func TestWatchdog_StreamCompletesOnCancel(t *testing.T) {
	w := circleai.NewDefaultSecurityWatchdog()
	ctx, cancel := context.WithCancel(context.Background())
	stream := w.StreamSignals(ctx)
	cancel()

	select {
	case _, ok := <-stream:
		if ok {
			// A late value is acceptable only if the channel then closes; drain.
			select {
			case _, ok2 := <-stream:
				if ok2 {
					t.Error("stream should close after cancel")
				}
			case <-time.After(time.Second):
				t.Error("stream did not close after cancel")
			}
		}
	case <-time.After(2 * time.Second):
		t.Error("stream did not close after cancel")
	}
}

File: rust/tests/security_watchdog_test.rs
//! security_watchdog_test.rs
//!
//! Ports `ISecurityWatchdog.cs` (`DefaultSecurityWatchdog`),
//! `IAnomalyEventDispatcher.cs`, `SecurityResponse.cs`, and
//! `SecurityCheckpoint.cs` graduated-response behaviour.

use std::sync::Arc;

use circle_ai::security::{
    confidence_band, AnomalyDispatchOutcome, AnomalySignal, DefaultAnomalyEventDispatcher,
    DefaultSecurityWatchdog, IAnomalyEventDispatcher, ISecurityWatchdog, SecurityCheckpoint,
    SecurityResponse, SecurityResponseKind, ThreatVector,
};

// ── DefaultSecurityWatchdog graduated policy ────────────────────────────────

#[test]
fn low_confidence_is_no_action() {
    let wd = DefaultSecurityWatchdog::new();
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.2, "mod", "low", None);
    let resp = wd.on_anomaly_detected(&sig, None);
    assert_eq!(resp.kind, SecurityResponseKind::NoAction);
    assert_eq!(resp.signal_id, sig.id);
    assert!(resp.applied_actions.is_empty());
    assert!(resp.restored_checkpoint.is_none());
}

#[test]
fn mid_confidence_rotates_keys() {
    let wd = DefaultSecurityWatchdog::new();
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.45, "mod", "mid", None);
    let resp = wd.on_anomaly_detected(&sig, None);
    assert_eq!(resp.kind, SecurityResponseKind::KeyRotation);
    assert!(resp.applied_actions.is_empty());
}

#[test]
fn boundary_030_is_key_rotation_not_no_action() {
    // Confidence == 0.30 is NOT < 0.30, so it escalates to rotation.
    let wd = DefaultSecurityWatchdog::new();
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.30, "mod", "edge", None);
    let resp = wd.on_anomaly_detected(&sig, None);
    assert_eq!(resp.kind, SecurityResponseKind::KeyRotation);
}

#[test]
fn boundary_060_is_key_rotation_not_composite() {
    // Confidence == 0.60 is NOT > 0.60, so it stays key-rotation.
    let wd = DefaultSecurityWatchdog::new();
    let sig = AnomalySignal::create(ThreatVector::ControlFlowDrift, 0.60, "mod", "edge", None);
    let resp = wd.on_anomaly_detected(&sig, None);
    assert_eq!(resp.kind, SecurityResponseKind::KeyRotation);
}

#[test]
fn high_confidence_is_composite_rotation_plus_mesh() {
    let wd = DefaultSecurityWatchdog::new();
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.92, "mod", "high", None);
    let resp = wd.on_anomaly_detected(&sig, None);
    assert_eq!(resp.kind, SecurityResponseKind::Composite);
    assert_eq!(
        resp.applied_actions,
        vec![
            SecurityResponseKind::KeyRotation,
            SecurityResponseKind::MeshIsolationSignal
        ]
    );
    // MemoryAnomaly is NOT high-severity -> no rollback even with checkpoint.
    assert!(resp.restored_checkpoint.is_none());
}

#[test]
fn high_confidence_high_severity_with_verified_checkpoint_adds_rollback() {
    let wd = DefaultSecurityWatchdog::new();
    let cp = SecurityCheckpoint::create("uhid-1", "CircleAI.Memory", b"trusted-state".to_vec());
    let sig = AnomalySignal::create(ThreatVector::StateCorruption, 0.95, "CircleAI.Memory", "hi", None);
    let resp = wd.on_anomaly_detected(&sig, Some(&cp));
    assert_eq!(resp.kind, SecurityResponseKind::Composite);
    assert!(resp
        .applied_actions
        .contains(&SecurityResponseKind::StateRollback));
    assert!(resp.restored_checkpoint.is_some());
    assert_eq!(resp.restored_checkpoint.unwrap().id, cp.id);
}

#[test]
fn high_severity_but_tampered_checkpoint_skips_rollback() {
    let wd = DefaultSecurityWatchdog::new();
    let mut cp = SecurityCheckpoint::create("uhid-1", "CircleAI.Memory", b"good".to_vec());
    // Tamper with the payload so verify() fails.
    cp.payload = b"evil".to_vec();
    assert!(!cp.verify());
    let sig = AnomalySignal::create(ThreatVector::NetworkPivot, 0.99, "mod", "hi", None);
    let resp = wd.on_anomaly_detected(&sig, Some(&cp));
    assert_eq!(resp.kind, SecurityResponseKind::Composite);
    assert!(!resp
        .applied_actions
        .contains(&SecurityResponseKind::StateRollback));
    assert!(resp.restored_checkpoint.is_none());
}

#[test]
fn stream_signals_drains_every_processed_signal() {
    let wd = DefaultSecurityWatchdog::new();
    let s1 = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.1, "m", "a", None);
    let s2 = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.9, "m", "b", None);
    wd.on_anomaly_detected(&s1, None);
    wd.on_anomaly_detected(&s2, None);
    let drained = wd.stream_signals();
    assert_eq!(drained.len(), 2);
    assert_eq!(drained[0].id, s1.id);
    assert_eq!(drained[1].id, s2.id);
    assert!(wd.stream_signals().is_empty(), "second drain empty");
}

#[test]
fn component_name_matches_reference() {
    assert_eq!(DefaultSecurityWatchdog::new().component_name(), "DefaultSecurityWatchdog");
}

#[test]
fn confidence_band_matches_reference_bands() {
    assert_eq!(confidence_band(0.0), "low");
    assert_eq!(confidence_band(0.29), "low");
    assert_eq!(confidence_band(0.30), "mid");
    assert_eq!(confidence_band(0.59), "mid");
    assert_eq!(confidence_band(0.60), "high");
    assert_eq!(confidence_band(1.0), "high");
}

// ── SecurityResponse factories ──────────────────────────────────────────────

#[test]
fn rollback_factory_records_checkpoint_and_description() {
    let cp = SecurityCheckpoint::create("uhid", "ModX", b"p".to_vec());
    let id = uuid_like();
    let resp = SecurityResponse::for_rollback(id, cp.clone());
    assert_eq!(resp.kind, SecurityResponseKind::StateRollback);
    assert_eq!(resp.restored_checkpoint.as_ref().unwrap().id, cp.id);
    assert!(resp.description.contains(&cp.id.to_string()));
    assert!(resp.description.contains("ModX"));
}

fn uuid_like() -> uuid::Uuid {
    uuid::Uuid::new_v4()
}

// ── DefaultAnomalyEventDispatcher ───────────────────────────────────────────

#[test]
fn dispatcher_forwards_qualifying_signal() {
    let wd: Arc<dyn ISecurityWatchdog> = Arc::new(DefaultSecurityWatchdog::new());
    let disp = DefaultAnomalyEventDispatcher::new(Arc::clone(&wd));
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.9, "m", "x", None);
    let r = disp.verify_and_dispatch(&sig, None, false);
    assert_eq!(r.outcome, AnomalyDispatchOutcome::Dispatched);
    assert!(r.response.is_some());
    assert_eq!(r.response.unwrap().kind, SecurityResponseKind::Composite);
}

#[test]
fn dispatcher_drops_below_threshold() {
    let wd: Arc<dyn ISecurityWatchdog> = Arc::new(DefaultSecurityWatchdog::new());
    let disp = DefaultAnomalyEventDispatcher::new(Arc::clone(&wd)); // default min 0.30
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.2, "m", "low", None);
    let r = disp.verify_and_dispatch(&sig, None, false);
    assert_eq!(r.outcome, AnomalyDispatchOutcome::BelowThreshold);
    assert!(r.response.is_none());
    // Watchdog never saw it.
    assert!(wd.stream_signals().is_empty());
}

#[test]
fn dispatcher_dedupes_by_id() {
    let wd: Arc<dyn ISecurityWatchdog> = Arc::new(DefaultSecurityWatchdog::new());
    let disp = DefaultAnomalyEventDispatcher::new(Arc::clone(&wd));
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.9, "m", "x", None);
    let r1 = disp.verify_and_dispatch(&sig, None, false);
    let r2 = disp.verify_and_dispatch(&sig, None, false);
    assert_eq!(r1.outcome, AnomalyDispatchOutcome::Dispatched);
    assert_eq!(r2.outcome, AnomalyDispatchOutcome::Duplicate);
    assert!(r2.response.is_none());
    assert_eq!(wd.stream_signals().len(), 1, "watchdog invoked once");
}

#[test]
fn dispatcher_reports_cancellation() {
    let wd: Arc<dyn ISecurityWatchdog> = Arc::new(DefaultSecurityWatchdog::new());
    let disp = DefaultAnomalyEventDispatcher::new(Arc::clone(&wd));
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.9, "m", "x", None);
    let r = disp.verify_and_dispatch(&sig, None, true);
    assert_eq!(r.outcome, AnomalyDispatchOutcome::Cancelled);
}

#[test]
fn dispatcher_custom_threshold_is_clamped() {
    let wd: Arc<dyn ISecurityWatchdog> = Arc::new(DefaultSecurityWatchdog::new());
    // Above-1 threshold clamps to 1.0, so 0.99 falls below.
    let disp = DefaultAnomalyEventDispatcher::with_minimum_confidence(Arc::clone(&wd), 2.0);
    assert_eq!(disp.minimum_confidence(), 1.0);
    let sig = AnomalySignal::create(ThreatVector::MemoryAnomaly, 0.99, "m", "x", None);
    assert_eq!(
        disp.verify_and_dispatch(&sig, None, false).outcome,
        AnomalyDispatchOutcome::BelowThreshold
    );
}
